File: java/src/main/java/cn/az/code/year2020/nov/BinaryTreeTilt.java
package cn.az.code.year2020.nov;

import cn.az.code.common.TreeNode;

/**
 * @author az
 * @since 11/08/20
 */
public class BinaryTreeTilt {

    int result = 0;

    public int findTilt(TreeNode root) {

        traverse(root);
        return result;
    }

    private void traverse(TreeNode root) {
        if (root == null) {
            return;
        }
        result += helper(root);
        traverse(root.left);
        traverse(root.right);
    }

    private int helper(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int left = count(node.left);
        int right = count(node.right);

        return Math.abs(left - right);
    }


    private int count(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int left = count(node.left);
        int right = count(node.right);
        return left + right + node.val;
    }

}

File: archive/cpp/2021/dec/binary-tree-tilt.cc
#include "common.h"

class Solution
{
public:
    int findTilt(TreeNode *root)
    {
        if (!root)
        {
            return 0;
        }
        int res = 0;
        post_order(root, res);

        return res;
    }

private:
    int post_order(TreeNode *root, int &res)
    {
        if (!root)
        {
            return 0;
        }

        int left_sum = post_order(root->left, res);

        int right_sum = post_order(root->right, res);

        res += abs(left_sum - right_sum);

        return left_sum + right_sum + root->val;
    }
};